In [1]:
#IPython is what you are using now to run the notebook
import IPython
print( "IPython version:      %6.6s (need at least 1.0)" % IPython.__version__)

# Numpy is a library for working with arrays and matrices
import numpy as np
print( "Numpy version:        %6.6s (need at least 1.7.1)" % np.__version__)

# SciPy implements many different numerical algorithms
import scipy as sp
print( "SciPy version:        %6.6s (need at least 0.12.0)" % sp.__version__)

# Pandas makes working with data tables easier
import pandas as pd
print( "Pandas version:       %6.6s (need at least 0.11.0)" % pd.__version__)

# Module for plotting
import matplotlib.pyplot as plt  
from pylab import *
print( "Mapltolib version:    %6.6s (need at least 1.2.1)" %
       matplotlib.__version__)
%matplotlib inline
# necessary for in-line graphics

# SciKit Learn implements several Machine Learning algorithms
import sklearn
print( "Scikit-Learn version: %6.6s (need at least 0.13.1)" %
       sklearn.__version__)
import os


# for certain system-related functions

from scipy import stats

import statsmodels.formula.api as smf


IPython version:       7.6.1 (need at least 1.0)
Numpy version:        1.16.4 (need at least 1.7.1)
SciPy version:         1.2.1 (need at least 0.12.0)
Pandas version:       0.24.2 (need at least 0.11.0)
Mapltolib version:     3.1.0 (need at least 1.2.1)
Scikit-Learn version: 0.21.2 (need at least 0.13.1)


In [2]:
#data = pd.read_csv("progresa-sample.csv", sep=",")
data = pd.read_csv("progresa-sample.csv")

In [3]:
data.shape

(77250, 21)

**Q1. First, learn about data.**

***Q1.1 Report summary statistics (mean, standard deviation, and number of missings) for all of the demo-
graphic variables in the dataset (i.e., everything except year, folnum, village). A central variable,
progresa is coded in a rather unintuitive way. Find it's actual coding scheme. Does this fit with the
documentation above?
Present these in a single table alphabetized by variable name. Do NOT simply expect the grader to
scroll through your output!***

In [4]:
data.columns

Index(['year', 'sex', 'indig', 'dist_sec', 'sc', 'grc', 'fam_n', 'min_dist',
       'dist_cap', 'poor', 'progresa', 'hohedu', 'hohwag', 'welfare_index',
       'hohsex', 'hohage', 'age', 'village', 'folnum', 'grc97', 'sc97'],
      dtype='object')

In [5]:
data.sc.isnull().sum()

8453

In [6]:
data['progresa'].unique()

array(['0', 'basal'], dtype=object)

In [7]:
data['poor'].unique()

array(['pobre', 'no pobre'], dtype=object)

In [8]:
"""Progresa column has two values- 'basal' for treatment and  0 for control. It means basal is put for those villages who received the subsidy.
And 0 is for those villages who did not receive the subsidy.It is dividing in tretment nad control groups.

"""

"Progresa column has two values- 'basal' for treatment and  0 for control. It means basal is put for those villages who received the subsidy.\nAnd 0 is for those villages who did not receive the subsidy.It is dividing in tretment nad control groups.\n\n"

In [9]:
"""poor also has two variables - pobre and no pobre. pobre is for those families which were categorised as poor and non pobre for those which were categorised non poor.
It is blocking the objects in treatment and control groups into two groups."""

'poor also has two variables - pobre and no pobre. pobre is for those families which were categorised as poor and non pobre for those which were categorised non poor.\nIt is blocking the objects in treatment and control groups into two groups.'

In [10]:
#We would change the values in Progressa- 1 for treatment(basal) and 0 for control
#We would also change the values in poor- 1 for poor and 0 for non poor

In [11]:
df = data.copy()

In [12]:
data.shape

(77250, 21)

In [13]:
df.progresa.replace(['0', 'basal'], [0, 1], inplace=True) #changing pregresa column values- 1 for treatment, 0 for control

In [14]:
df.poor.replace(['pobre', 'no pobre'], [1, 0], inplace=True) #changing poor column values- 1 for poor, 0 for non poor

In [15]:
#creating an empty data frame for summary statistics- variable, mean, standard deviation, number of missing values
summary = pd.DataFrame(columns=['variable', 'mean', 'standard_deviation','num_missingvalues'])

In [16]:
#Creating a list of variables for which we have to find summary, this has all the variables except year, folnum, village
variables = ['sex', 'indig', 'dist_sec', 'sc', 'grc', 'fam_n', 'min_dist',
       'dist_cap', 'poor', 'progresa', 'hohedu', 'hohwag', 'welfare_index',
       'hohsex', 'hohage', 'age',  'grc97', 'sc97']

In [17]:
for variable in variables:
    mean = df[variable].mean()
    stdev = df[variable].std()
    num_of_missing_values = df[variable].isnull().sum()
    summary = summary.append({'variable' : variable , 'mean' : mean, 'standard_deviation':stdev, 'num_missingvalues':num_of_missing_values } , ignore_index=True)
    
    
    

In [18]:
summary.sort_values(by=['variable'])

variable        mean  standard_deviation num_missingvalues
15            age   11.366460            3.167744                 0
7        dist_cap  147.674452           76.063134                 0
2        dist_sec    2.418910            2.234109                 0
5           fam_n    7.215715            2.352900                 0
4             grc    3.963537            2.499063              6549
16          grc97    3.705372            2.572387                 0
14         hohage   44.436717           11.620372                10
10         hohedu    2.768104            2.656106                 0
13         hohsex    0.925185            0.263095                20
11         hohwag  586.985312          788.133664                 0
1           indig    0.298324            0.457525               300
6        min_dist  103.447520           42.089441                 0
8            poor    0.846498            0.360473                 0
9        progresa    0.615663            0.486441                 0
3              sc    0.819818            0.384342              8453
17           sc97    0.813922            0.389172              3872
0             sex    0.512211            0.499854                24
12  welfare_index  690.346564          139.491130               210

In [19]:
"""This table gives the mean, standard deviation and number of missing values of all the columns.
"""

'This table gives the mean, standard deviation and number of missing values of all the columns.\n'

**Q1.2- Now let's investigate the differences in baseline. Are the baseline (1997) demographic characteristics for
the poor different in treatment and control villages?**

***Q 1.2- 1- Use t-test to determine whether there is a statistically significant difference in the average
values of each of the variables in the dataset. Focus only on the data from 1997 for poor.***

We have to do t test for different values of the poors of treatment and control villages. We are only talking about 1997. The progresa column has 1 for treatment villages and 0 for control villages.


In [20]:
#We would select only those rows which have poor family and year 1997
poor_1997 = df[(df['poor'] == 1) & (df['year'] == 97)]

In [21]:
#Question 1 told us to ignore year, folnum, village columns, so we would remove them. We would also remove year and poor columns.
poor_1997 = poor_1997.drop(['year', 'folnum', 'village', 'poor'], axis = 1)


In [22]:
#We would not make two subsets of data, treatment(progresa = 1) and control(progresa = 0)
#We also drop the pregresa columns

treatment = poor_1997[poor_1997['progresa'] == 1].drop(['progresa'],axis =1)

control = poor_1997[poor_1997['progresa'] == 0].drop(['progresa'],axis =1)

In [23]:
#creating an empty data frame to hold values from dictionary
t_tests = pd.DataFrame(columns=['variable', 'avg_treatment_village', 'avg_control_village','difference', 'p_value'])

In [24]:
treatment.columns == control.columns

#This tells that the columns in treatment and control data frames are not different

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True])

In [25]:
from scipy import stats
for column in treatment.columns:
    
    
    #variables_dictionary["variable"] = column
    
    avg_treatment = round(treatment[column].dropna().mean(),6) #finding mean of that column for treatment villages
    
    
    avg_control = round(control[column].dropna().mean(),6) #finding average of that column for control voillages
    
    
    diff = avg_treatment - avg_control #finding the difference
    
    
    stat1 = stats.ttest_ind(treatment[column].dropna(),control[column].dropna(), equal_var = False) #doing the t test
    p_value = stat1[1]
    
    
    t_tests = t_tests.append({'variable':column, 'avg_treatment_village': avg_treatment, 'avg_control_village': avg_control,'difference': diff, 'p_value':p_value } , ignore_index=True)
    
    
    

***Q1.2.2- Present your results in a single table with the following columns and 14 (or so) rows:***

In [26]:
t_tests

#The t_tests data frame has all the variables, its average for treatment villages, its average for control villages, difference in the averages and p_value from t -test

variable  avg_treatment_village  avg_control_village  difference  \
0             sex               0.519317             0.505052    0.014265   
1           indig               0.325986             0.332207   -0.006221   
2        dist_sec               2.453122             2.507662   -0.054540   
3              sc               0.822697             0.815186    0.007511   
4             grc               3.531599             3.543050   -0.011451   
5           fam_n               7.281327             7.302469   -0.021142   
6        min_dist             107.152915           103.237854    3.915061   
7        dist_cap             150.829074           153.769730   -2.940656   
8          hohedu               2.663139             2.590348    0.072791   
9          hohwag             544.339544           573.163558  -28.824014   
10  welfare_index             655.428377           659.579100   -4.150723   
11         hohsex               0.924656             0.922947    0.001709   
12         hohage              43.648828            44.276918   -0.628090   
13            age              10.716991            10.742023   -0.025032   
14          grc97               3.531599             3.543050   -0.011451   
15           sc97               0.822697             0.815186    0.007511   

         p_value  
0   1.220744e-02  
1   2.459021e-01  
2   4.266282e-02  
3   9.646120e-02  
4   6.895228e-01  
5   4.289667e-01  
6   7.055795e-16  
7   1.146482e-03  
8   1.038219e-02  
9   3.287285e-04  
10  1.531678e-03  
11  5.721253e-01  
12  2.259461e-06  
13  4.783633e-01  
14  6.895228e-01  
15  9.646120e-02

The estimators like sex,indig,dist_sec,min_dist,hoage,age have statistically significant value even before the treatment. we did a t test and their p-value is less than 0.05

***Q1.2.3- Do you find any statistically significant differences between treatment and control villages as
baseline?***

The estimators like sex,indig,dist_sec,min_dist,hoage,age have statistically significant value even before the treatment. we did a t test and their p-value is less than 0.05. So these variables between treatment and control groups are statistically significantly different at the baseline.

***Q1.2.4 Why does it matter if there are differences at baseline?***

If these predictors are different at baseline, so our causal inference is supicious. It has an inherent bias. The change that we observe might not be just because of the experiment. The intial differences would have an impact on the experiment.
Our main assumption is that external factors do not affect our treatment or control groups. If there is difference at the baseline, which means the sampling which we did has biased samples.

***Q1.2.5 What does this imply about how to measure the impact of the treatment?***

It implies that the causal inference which we do may not be totally correct. The indicators have a difference at baseline, this means that even afyter randomisation there would be bias and the effect of exyternal factors in our tests.

**2.1 First, implement the before-after estimator. Compare the schooling rate of poor households in progresa
villages before (i.e. 1997) and after (i.e. 1998) the program.**

***Q2.1.1 compute the estimator by just comparing the average schooling rates for these villages.***

In [27]:
#We have to compute before(1997)- after(1998) for Progresa(treatmet) villages. We only have to compute for poor househilds as only they received the treatment.

#We have to clean the dataset and apply filter for getting required rows, we only select rows having progresa = 1 and poor = 1
before_after = df[(df['progresa'] ==1) & (df['poor'] ==1)]

In [55]:
df.shape

(77250, 21)

In [56]:
#We would group by year and find mean of sc

sc_rate_mean = before_after[['sc','year']].groupby(['year']).mean()

#sc_rate_mean = progresa_before_after.groupby('year').sc.mean()

sc_rate_mean

sc
year          
97    0.822697
98    0.846479

In [58]:
increase_in_schooling = 0.846479 - 0.822697
increase_in_schooling

0.02378199999999997

In [62]:

#We find that the before schooling rate was 0.822697; the after schooling rate was 0.846479
#The difference in means = 0.02378
#We can say that there was 2.378% increasing in schooling rate after the treatment.

***Q2.1.2 - now re-compute the estimator using linear regression, and individual schooling rates. Do not
include other regressors.***

In [61]:
#We would fit a linear regression model for sc as response variable and year as predictor variable to understand their relationship.
m = smf.ols(formula = 'sc ~ year', data= before_after)
r = m.fit()
r.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     sc   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     36.84
Date:                Thu, 06 Feb 2020   Prob (F-statistic):           1.30e-09
Time:                        19:41:24   Log-Likelihood:                -15557.
No. Observations:               36175   AIC:                         3.112e+04
Df Residuals:                   36173   BIC:                         3.113e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.4842      0.382     -3.886      0.000      -2.233      -0.736
year           0.0238      0.004      6.069      0.000       0.016       0.031
==============================================================================
Omnibus:                    10294.733   Durbin-Watson:                   1.394
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            21582.139
Skew:                          -1.791   Prob(JB):                         0.00
Kurtosis:                       4.217   Cond. No.                     1.90e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.9e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [63]:
#The p value is almost 0 for year, which tells it is statistically signioficant. 
#The value of coefficient of year = 0.0238, it is equal to what we had got earlier from finding the mean. It is the difference over time.
#We would observe an increse of 0.0238 units in schooling rate(sc) for 1 unit increase in year.

***Q2.1.3 - Finally, estimate a multiple regression model that includes other covariates.***

In [34]:
before_after.columns

Index(['year', 'sex', 'indig', 'dist_sec', 'sc', 'grc', 'fam_n', 'min_dist',
       'dist_cap', 'poor', 'progresa', 'hohedu', 'hohwag', 'welfare_index',
       'hohsex', 'hohage', 'age', 'village', 'folnum', 'grc97', 'sc97'],
      dtype='object')

In [64]:
#We would fit a multilunear regression model to check how sc depends on other variables
multiple = smf.ols(formula = 'sc ~ year + sex+ indig+ dist_sec+ grc+ fam_n+ min_dist+ dist_cap+ hohedu+ hohwag+ welfare_index+hohsex+ hohage+ age', data= before_after)
r1 = multiple.fit()
r1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     sc   R-squared:                       0.301
Model:                            OLS   Adj. R-squared:                  0.301
Method:                 Least Squares   F-statistic:                     1108.
Date:                Thu, 06 Feb 2020   Prob (F-statistic):               0.00
Time:                        19:42:44   Log-Likelihood:                -9044.6
No. Observations:               35972   AIC:                         1.812e+04
Df Residuals:                   35957   BIC:                         1.825e+04
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept        -3.5391      0.323    -10.971      0.000      -4.171      -2.907
year              0.0530      0.003     16.024      0.000       0.046       0.059
sex               0.0361      0.003     10.998      0.000       0.030       0.043
indig             0.0300      0.004      7.382      0.000       0.022       0.038
dist_sec         -0.0067      0.001     -8.274      0.000      -0.008      -0.005
grc               0.0473      0.001     37.148      0.000       0.045       0.050
fam_n             0.0012      0.001      1.607      0.108      -0.000       0.003
min_dist          0.0003   5.44e-05      5.413      0.000       0.000       0.000
dist_cap          0.0002   3.29e-05      7.338      0.000       0.000       0.000
hohedu            0.0031      0.001      4.183      0.000       0.002       0.005
hohwag         2.023e-06   2.41e-06      0.839      0.401    -2.7e-06    6.75e-06
welfare_index -3.586e-05   1.57e-05     -2.284      0.022   -6.66e-05   -5.09e-06
hohsex           -0.0028      0.007     -0.432      0.666      -0.016       0.010
hohage        -8.056e-05      0.000     -0.476      0.634      -0.000       0.000
age              -0.0950      0.001    -92.975      0.000      -0.097      -0.093
==============================================================================
Omnibus:                     4343.381   Durbin-Watson:                   1.496
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             6103.523
Skew:                          -0.957   Prob(JB):                         0.00
Kurtosis:                       3.639   Cond. No.                     2.01e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.01e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

***Q2.1.4 - compare all the estimators. Are your estimates statistically significant? What do they suggest
about the efficacy of the progresa program?***


The statistically significant predictors are year, sex, indig, dist_sec, grc, min_dist, dist_cap,hohedu,welfare_index,age. When we have taken all the predictors, the coefficient of year has increased from 0.0238 to 0.0530.
It means that when we consider the effect of other predictors, it enhances the effect of treatment over time.
The progresa program is effective overall.


**Q2.2 Now let's implement the cross-sectional estimator. Proceed along the same lines as what you did above.**

***Q2.2.1 Begin by estimating the impact of Progresa by compring the average enrollment rate among
poor households in the treatment villages and the average enrollment rate among poor households
in the control villages. What do you find?***

The average difference in outcomes between treated and non-treated individuals is due to the treatment, and only due to the treatment.
So we would check in 1998 (after), the average enrollment rate in poor households of df(treatment) and non df(control).

In [66]:
# We would only check for after treatment and in poor households. We would filter the data for year = 1998 and poor =1.
#This would enalbe us to compare schooling rate (sc) between poor df(treatment) and poor non df(control) households.
cross_section = df[(df['year'] ==98) & (df['poor'] ==1)]


In [67]:
#We would group by progresa(treatment) and find mean of schooling rate(sc).

sc_rate_mean_cross_section = cross_section[['sc','progresa']].groupby(['progresa']).mean()

sc_rate_mean_cross_section


sc
progresa          
0         0.807637
1         0.846479

In [68]:
difference_cross_section = 0.846479 - 0.807637
difference_cross_section

0.03884199999999993

In [72]:
""" The difference in the schooling rate(sc) of poor households of df(treatment) and non progresa(control) group is 0.0388 which is 3.88%.
"""

' The difference in the schooling rate(sc) of poor households of df(treatment) and non progresa(control) group is 0.0388 which is 3.88%.\n'

***Q2.2.2 Now repeat the estimator using simple regression.***

In [73]:
cs_m = smf.ols(formula = 'sc ~ progresa', data= cross_section)
cs_m_2 = cs_m.fit()
cs_m_2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     sc   R-squared:                       0.003
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     69.87
Date:                Thu, 06 Feb 2020   Prob (F-statistic):           6.64e-17
Time:                        19:49:17   Log-Likelihood:                -11926.
No. Observations:               27450   AIC:                         2.386e+04
Df Residuals:                   27448   BIC:                         2.387e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.8076      0.004    220.676      0.000       0.800       0.815
progresa       0.0388      0.005      8.359      0.000       0.030       0.048
==============================================================================
Omnibus:                     7638.939   Durbin-Watson:                   1.734
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            15767.534
Skew:                          -1.767   Prob(JB):                         0.00
Kurtosis:                       4.140   Cond. No.                         3.01
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""


The p value of progresa variable is almost 0, which means it is statistically significant. 
The intercept is 0.8076. The schooling rate among non progresa households was also 0.8076

The progresa coeficient is 0.0388. So the slope is 0.0388 which is eqaul to what we found above from difference of mean schooling rates of progresa and non progresa groups.

We would observe an increse of 0.0388 units in schooling rate(sc) for 1 unit increase in progresa variable.



***Q2.2.3 Third, use multiple regression to get the same estimate.***

In [42]:
cross_section.columns

Index(['year', 'sex', 'indig', 'dist_sec', 'sc', 'grc', 'fam_n', 'min_dist',
       'dist_cap', 'poor', 'progresa', 'hohedu', 'hohwag', 'welfare_index',
       'hohsex', 'hohage', 'age', 'village', 'folnum', 'grc97', 'sc97'],
      dtype='object')

In [74]:
multiple_cs = smf.ols(formula = 'sc ~ progresa + sex+ indig+ dist_sec+ grc+ fam_n+ min_dist+ dist_cap+ hohedu+ hohwag+ welfare_index+hohsex+ hohage+ age', data= cross_section)
r1_cs = multiple_cs.fit()
r1_cs.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     sc   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                     847.0
Date:                Thu, 06 Feb 2020   Prob (F-statistic):               0.00
Time:                        19:49:39   Log-Likelihood:                -6928.7
No. Observations:               27200   AIC:                         1.389e+04
Df Residuals:                   27185   BIC:                         1.401e+04
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         1.6540      0.020     84.276      0.000       1.616       1.692
progresa          0.0319      0.004      8.140      0.000       0.024       0.040
sex               0.0336      0.004      8.855      0.000       0.026       0.041
indig             0.0272      0.005      5.852      0.000       0.018       0.036
dist_sec         -0.0074      0.001     -8.537      0.000      -0.009      -0.006
grc               0.0510      0.001     34.772      0.000       0.048       0.054
fam_n             0.0015      0.001      1.724      0.085      -0.000       0.003
min_dist          0.0004   6.22e-05      5.653      0.000       0.000       0.000
dist_cap          0.0002   3.62e-05      5.247      0.000       0.000       0.000
hohedu            0.0041      0.001      4.790      0.000       0.002       0.006
hohwag        -1.515e-06   2.77e-06     -0.547      0.584   -6.94e-06    3.91e-06
welfare_index -1.809e-05   1.79e-05     -1.011      0.312   -5.32e-05     1.7e-05
hohsex            0.0051      0.008      0.672      0.502      -0.010       0.020
hohage           -0.0003      0.000     -1.571      0.116      -0.001    7.49e-05
age              -0.1000      0.001    -83.566      0.000      -0.102      -0.098
==============================================================================
Omnibus:                     3250.561   Durbin-Watson:                   1.733
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4554.560
Skew:                          -0.949   Prob(JB):                         0.00
Kurtosis:                       3.645   Cond. No.                     1.08e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.08e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

***Q2.2.4 Finally, as above, compare your three estimators. What do you find? Are the effects statistically significant?***

The statistically significant predictors are progresa, sex, indig, dist_sec, grc,min_dist,dist_cap,hohedu,age. When we have taken all the predictors, the coefficient of progresa has decreased from 0.0388
 to 0.0319.
It means that when we consider the effect of other predictors, it decreases the effect of treatment over time.
Still the progresa program is effective in increasing the schooling rate overall.


**Q2.3 Now we are ready for DiD estimator. Proceed along the same lines as above.**


***Q2.3.1 Start with the simple table. However, DiD requires 4-way comparison. So compare the average enrollment rate among poor households in the treatment villages and the average enrollment rate
among poor households in the control villages, both 1997 and 1998. What do you find?***

In [75]:
#We have to find the average enrollment rates among poor households of treatment and control villages for 1997 and 1998. 
#So we only select the rows having poor = 1.
dif_in_dif = df[ df['poor'] ==1]

In [76]:
#We have got poor households df(treatment) and non df(control) groups.
#We would now group by df and year; and find mean of sc(schooling rate).
sc_rate_mean_diff_in_diff = dif_in_dif[['sc','progresa','year']].groupby(['progresa','year']).mean()
sc_rate_mean_diff_in_diff

sc
progresa year          
0        97    0.815186
         98    0.807637
1        97    0.822697
         98    0.846479

In [77]:
diff_in_1997 = 0.822697 - 0.815186 #difference of treatment and control in 1997
diff_in_1998 = 0.846479 - 0.807637 #difference of treatment and control in 1998

In [78]:
diff_in_1997

0.007511000000000045

In [79]:
diff_in_1998

0.03884199999999993

In [89]:
difference_in_difference = diff_in_1998 - diff_in_1997
difference_in_difference
#It is the difference in difference

0.03133099999999989

In 1997 the difference in poor households of df(treatment) and non progresa(control) groups was 0.007511000000000045.

In 1998 the difference in poor households of df(treatment) and non progresa(control) groups was 0.03884199999999993.

The difference in difference = 0.03884199999999993 - 0.007511000000000045 = 0.0313

***Q2.3.2- Now repeat the estimator using simple regression.***

In [93]:
#We would fit a simple regression model for schooling rate dependent on the interaction of progresa and year

did_m = smf.ols(formula = 'sc ~ progresa*year', data= dif_in_dif)
did_m_2 = did_m.fit()
did_m_2.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     sc   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     28.31
Date:                Thu, 06 Feb 2020   Prob (F-statistic):           2.76e-18
Time:                        20:34:50   Log-Likelihood:                -26242.
No. Observations:               58372   AIC:                         5.249e+04
Df Residuals:                   58368   BIC:                         5.253e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         1.5474      0.497      3.112      0.002       0.573       2.522
progresa         -3.0316      0.632     -4.800      0.000      -4.270      -1.794
year             -0.0075      0.005     -1.480      0.139      -0.018       0.002
progresa:year     0.0313      0.006      4.835      0.000       0.019       0.044
==============================================================================
Omnibus:                    15346.988   Durbin-Watson:                   1.397
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            30608.651
Skew:                          -1.711   Prob(JB):                         0.00
Kurtosis:                       3.937   Cond. No.                     5.72e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.72e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

We see that the intercept is 1.5474. The slope of progresa is -3.0316. The slope of year is -0.0075.
The interaction effect of progresa and year is 0.0313 which is equal to the difference in difference we found above.
So progresa treatment alone or year alone is decreasing the schooling rate, but when we consider their interaction effect, it increases the schooling rate.



***Q2.3.3- And as above, use multiple regression to get the same estimate.***

In [51]:
dif_in_dif.columns

Index(['year', 'sex', 'indig', 'dist_sec', 'sc', 'grc', 'fam_n', 'min_dist',
       'dist_cap', 'poor', 'progresa', 'hohedu', 'hohwag', 'welfare_index',
       'hohsex', 'hohage', 'age', 'village', 'folnum', 'grc97', 'sc97'],
      dtype='object')

In [91]:
#We would fit a multiple regression model for schooling rate dependent on the interaction of progresa and year; including other variables

did_multiple = smf.ols(formula = 'sc ~ progresa*year+sex+indig+ dist_sec+ grc+fam_n+min_dist+dist_cap+poor+progresa+ hohedu + hohwag + welfare_index+ hohsex+ hohage+ age', data= dif_in_dif)
did_multiple_2 = did_multiple.fit()
did_multiple_2.summary() 


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     sc   R-squared:                       0.311
Model:                            OLS   Adj. R-squared:                  0.310
Method:                 Least Squares   F-statistic:                     1631.
Date:                Thu, 06 Feb 2020   Prob (F-statistic):               0.00
Time:                        20:34:30   Log-Likelihood:                -15320.
No. Observations:               57938   AIC:                         3.067e+04
Df Residuals:                   57921   BIC:                         3.083e+04
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept        -0.4068      0.208     -1.954      0.051      -0.815       0.001
progresa         -2.8250      0.527     -5.358      0.000      -3.858      -1.792
year              0.0251      0.004      5.866      0.000       0.017       0.033
progresa:year     0.0292      0.005      5.390      0.000       0.019       0.040
sex               0.0367      0.003     14.006      0.000       0.032       0.042
indig             0.0317      0.003      9.854      0.000       0.025       0.038
dist_sec         -0.0065      0.001    -10.589      0.000      -0.008      -0.005
grc               0.0486      0.001     48.071      0.000       0.047       0.051
fam_n             0.0009      0.001      1.505      0.132      -0.000       0.002
min_dist          0.0003   4.31e-05      7.579      0.000       0.000       0.000
dist_cap          0.0002   2.51e-05      7.329      0.000       0.000       0.000
poor             -0.4068      0.208     -1.954      0.051      -0.815       0.001
hohedu            0.0037      0.001      6.225      0.000       0.003       0.005
hohwag          4.95e-07    1.9e-06      0.260      0.795   -3.24e-06    4.23e-06
welfare_index -1.631e-05   1.24e-05     -1.318      0.188   -4.06e-05    7.95e-06
hohsex           -0.0017      0.005     -0.324      0.746      -0.012       0.008
hohage        -4.707e-05      0.000     -0.354      0.723      -0.000       0.000
age              -0.0983      0.001   -120.910      0.000      -0.100      -0.097
==============================================================================
Omnibus:                     6083.138   Durbin-Watson:                   1.514
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             8175.414
Skew:                          -0.887   Prob(JB):                         0.00
Kurtosis:                       3.488   Cond. No.                     6.96e+18
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.29e-27. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

When we fit the multiple regression model, the coefficient of interaction term of year:progresa has reduced from 0.0313 to 0.0292. It means that the interaction term has less impact in presence of other variables.

***Q2.3.4 - Finally, as above, compare your three estimators. What do you find? Are the effects statistically significant?***

We do 4 way comparisons in difference in difference estimators.
In 2.3.1) we had initially applied filter to get only poor households. Those poor households with progresa = 1 is our treatment group. Those households with progresa = 0 is our control group. We grouped by progresa and year; then took mean sc for 1997 and 1998. This gave us a table for schooling rates for treatment and control groups in 1997 in 1998.
We took out the difference of schooling rates of treatment and control groups in 1997 and 1998. We subtracted it to get difference in difference estimator which was 0.0313

In 2.3.2 we did a simple .linear regression of sc over the interaction of year and progresa. We found the interaction term of progresa and year to be statistically significant. The coefficient of the interaction effect was 0.0313 which is equal to the difference in difference we got in 2.3.1

In 2.3.3 we did a multiple regression of sc over the interaction of progresa and year, and also added the other variables.
We find the following terms statistically significant: interaction term of progresa and year, progresa,year,progresa:year,sex,indig,dist_sec,grc,min_dist,dist_cap,poor,hohedu,age.
The coefficient of interaction term of year and progresa reduced to 0.0292. It means that the other variables are reducing the treatment effect over time

**Q2.4 Now you have used three estimators to assess the effect of Progresa program.**

***Q2.4.1 List the identifying assumptions (counterfactual assumptions) behind all three models. Which ones do you find more/less plausible?***

In the before after estimator, we compare only the treatment group before and after the treatment. Our underlying assumption is that the error terms or outside influence should not change with time, which means there should be no unobserved trend. So if there is an increase in enrollment between 1997 and 1998 it is because of treatment alone. This cannot be correct unless there is no time differences without treatment. However, it is not plausible as even if we were not doing any experiment, the schooling rate of children differs over time becuase of external of family reasons.

In the cross section estimator, we try to infer the average difference in outcomes between treated and non-treated individuals is due to the treatment, and only due to the treatment. So we measured in 1998(after) the difference of schoolng rates between treatment and control groups. The condition of mean independence is the underlying assumption. It means that the error term or mean effect of outside factors is the same for treated and non treated groups. I do not think it is fully plausible because households in different gropus would have gone through several different conditions. However randomisation helps to make this assumption stronger.

Difference in Difference compares the time trend for the treated group and nontreated group, or what amounts to the same, it compares treated-nontreated differences before and after the treatment. The underlying assumption is that without the treatment(progresa) experiment, the change in schooling rates of treatment and control groups would have been equal. It is difficult to assume it is fully true as we do not know how correctly was the data collected and how different different familoes would have gone through different situations. However randomisation would help to make this assumption stronger.

***Q2.4.2 Compare the estimates of all three models. Do your analysis suggest that progresa program
had a positive impact on schooling rates?***

We did before after test to compare the treated before the treatment and after the treatment. We found that the mean schooling rate in treatment group had increased from 1997 to 1998. Our linear regression models also told that year was statistically significant. The coefficient for year was 0.02378 in simple linear regression and 0.0530 in multiple linear egression.This means that the schooling rate did actually increase obver time in the treatment group. This told how the schooling rate in treated group would change with the change in time only because of treatment.

We did cross section to find the average difference in outcomes between treated and non-treated individuals is due to the treatment, and only due to the treatment. We found that the average schooling rate in treatment group was higher than control group. The coefficient of progresa in simple linear regression was 0.0388 and multiple linear regression was 0.0319. The progressa variable was also statistically significant. This tells how the schooling rate was different between the treated and control groups.

We then did the difference in difference test. We found the difference of schooling rate of treated and control groups in 1997. We also found the difference in schooling rate of treated and control groups in 1998. We the took the difference of these differences and found that the schooling rtae was higher in treated groups. We did simple and multiple linear regressions and that also told that the interaction term of progresa and year was statisticaly significant.

All of these tests tell that the Progresa program was a success and the schooling rate in treated village was higher than non treated group.